src/main/kotlin/ru/tretikoff/words/Word.kt

```diff
@@ -0,0 +1,35 @@
+package ru.tretikoff.words
+
+import java.util.logging.Logger
+
+enum class QuotationType {
+    None,
+    SingleQuoted,
+    DoubleQuoted,
+}
+
+private val logger = Logger.getLogger(Word::class.java.name)
+
+/**
+ * An abstraction of word token in the bash syntax
+ */
+class Word(var value: String, var quotationType: QuotationType = QuotationType.None) {
+    fun substituteVariables(variables: Map<String, String>) {
+        logger.finest("sadg")
+        while (true) {
+            val regex = "\\$\\w+".toRegex()
+            val match = regex.find(value) ?: return
+            val replacement = variables[match.value.drop(1)] ?: ""
+            logger.finest("replacing ${match.value} with $replacement in $value")
+            value = value.replace(regex, replacement)
```

Comment:
```suggestion
            value = value.replaceRange(match.range, replacement)
```
Из-за этого у вас заменяется сразу всё, что попадает под регэсп, на значение первой переменной.